pkg/util/push/request.go

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type Request struct {
+	cleanups []func()
+
+	httpReq                      *http.Request
+	maxMessageSize               int
+	allowSkipLabelNameValidation bool
+	parser                       ParserFunc
+	parseRequestOnce             sync.Once
+	parsedRequest                *mimirpb.WriteRequest
+	parseError                   error
+}
+
+func (r *Request) parseRequest(ctx context.Context) {
+	bufHolder := bufferPool.Get().(*bufHolder)
+	var req mimirpb.PreallocWriteRequest
+	buf, err := r.parser(ctx, r.httpReq, r.maxMessageSize, bufHolder.buf, &req)
+	buf = buf[:cap(buf)]
+	if err != nil {
+		// Check for httpgrpc error, default to client error if parsing failed
+		if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
+			err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+		}
+
+		bufferPool.Put(bufHolder)
+		r.parseError = err
+		return
+	}
+	// If decoding allocated a bigger buffer, put that one back in the pool.
+	if len(buf) > len(bufHolder.buf) {
+		bufHolder.buf = buf
+	}
+
+	r.AddCleanup(func() {
+		mimirpb.ReuseSlice(req.Timeseries)
+		bufferPool.Put(bufHolder)
+	})
+
+	if r.allowSkipLabelNameValidation {
+		req.SkipLabelNameValidation = req.SkipLabelNameValidation && r.httpReq.Header.Get(SkipLabelNameValidationHeader) == "true"
+	} else {
+		req.SkipLabelNameValidation = false
+	}
+
+	if req.Source == 0 {
+		req.Source = mimirpb.API
+	}
+
+	r.parsedRequest = &req.WriteRequest
+}
+
+// WriteRequest parses the body of the push request and caches it.
+// Subsequent calls to WriteRequest return the same results.
+func (r *Request) WriteRequest(ctx context.Context) (*mimirpb.WriteRequest, error) {
+	r.parseRequestOnce.Do(func() {
+		r.parseRequest(ctx)
+	})
+	return r.parsedRequest, r.parseError
+}
+
+// AddCleanup adds a function that will be called once CleanUp is called.
+func (r *Request) AddCleanup(f func()) {
+	r.cleanups = append(r.cleanups, f)
+}
+
+// CleanUp calls all added cleanups.
+func (r *Request) CleanUp() {
+	for _, f := range r.cleanups {
```

Comment:
I could only find this which changes the cleanup and it invokes the previous cleanups before doing its own cleanup

https://github.com/grafana/mimir/blob/3ecb9dcf76a7c6976f022d8a05e983ac65d48aae/pkg/distributor/distributor.go#L855-L859

Reply:
I am not sure which is a better sematic. Maybe calling the first cleanups last mimics defer semantics, so we want to be like that? But not sure what to do with the case ☝️ where we previously did the opposite.

Reply:
Even if today the order doesn't matter, I think it would be a preferred behaviour treating the cleanup function as a stack, so calling them in reverse order (looks more future proof to me).

Reply:
I pushed a change to do stack semantics in https://github.com/grafana/mimir/commit/e6ad4508ffa5260384eab490297bb39bc05db44d